lib/mixlib/install/product.rb

```diff
@@ -0,0 +1,264 @@
+                value
+              else
+                value.call(version_for(version))
+              end
+            else
+              instance_variable_set("@#{prop}".to_sym, prop_string)
+            end
+          else
+            raise "Not both at the same time." if !prop_string.nil?
+            instance_variable_set("@#{prop}".to_sym, block)
+          end
+        end
+      end
+
+      #
+      # Sets or retrieves the version for the product. This is used later
+      # when we are reading the value of a property if a Proc is specified
+      #
+      def version(value = nil)
+        if value.nil?
+          @version
+        else
+          @version = value
+        end
+      end
+
+      #
+      # Helper method to convert versions from String to Mixlib::Version
+      #
+      # @param [String] version_string
+      #   value to be set in String form
+      #
+      # @return [Mixlib::Version]
+      def version_for(version_string)
+        Mixlib::Versioning.parse(version_string)
+      end
+    end
+
+    class ProductMatrix
+      def initialize(&block)
+        @product_map = {}
+        instance_eval(&block)
+      end
+
+      #
+      # The only DSL method of this class. It creates a Product with given
+      # `key` and stores it.
+      #
+      def product(key, &block)
+        @product_map[key] = Product.new(&block)
+      end
+
+      #
+      # Fetches the keys of available products.
+      #
+      # @return Array[String] of keys
+      def products
+        @product_map.keys
+      end
+
+      #
+      # Looks up a product and sets version on it to be used later by the
+      # Product.
+      #
+      # @param [String] key
+      #   Lookup key of the product.
+      # @param [String] version
+      #   Version to be set for the product. By default version is set to :latest
+      #
+      # @return [Product]
+      def lookup(key, version = :latest)
+        product = @product_map[key]
+        # We set the lookup version for the product to a very high number in
+        # order to mimic :latest so that one does not need to handle this
+        # symbol explicitly when constructing logic based on version numbers.
+        version = "1000.1000.1000" if version.to_sym == :latest
+        product.version(version)
+        product
+      end
+    end
+  end
+end
+
+#
+# If you are making a change to PRODUCT_MATRIX, please make sure
+# you run `bundle exec rake matrix` at the home of this repository
+# to update PRODUCT_MATRIX.md.
+#
+PRODUCT_MATRIX = Mixlib::Install::ProductMatrix.new do
+  product "chef" do
+    product_name "Chef Client"
+    package_name "chef"
+  end
+
+  product "chefdk" do
+    product_name "Chef Development Kit"
+    package_name "chefdk"
+  end
+
+  product "chef-server" do
+    product_name "Chef Server"
+    package_name do |v|
+      if (v < version_for("12.0.0")) && (v > version_for("11.0.0"))
+        "chef-server"
+      else
+        "chef-server-core"
+      end
```

Comment:
Comment from that conditional block in the chef-ingredient code:

```
      # We want to validate that we're getting a version that is valid
      # for the Chef Server. However, since the default is 0.0.0,
      # implying latest, we need to additionally ensure that the
      # bottom version is something valid. If we don't have the check
      # in the elsif, it will say that 0.0.0 is not a valid version.
```